src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -57,117 +63,178 @@
-                this.streams[i] = stream;
-                this.streamFiles[i] = bclFile;
-                byteBuffer.clear();
-                headerBuffer.clear();
+            if (cycleData[totalCycleCount].tileInfo == null) {
+                throw new PicardException("Could not find tile " + cycleData[totalCycleCount].tileInfo);
             }
 
+            for (int outputLength : outputLengths) {
+                for (int cycle = 0; cycle < outputLength; cycle++) {
+                    CycleData currentCycleData = cycleData[totalCycleCount];
+                    try {
+                        if (cachedTile[totalCycleCount] == null) {
+                            if (!cachedFilter.containsKey(cycleData[totalCycleCount].tileInfo.tileNum)) {
+                                cacheFilterAndLocs(cycleData[totalCycleCount].tileInfo, locs);
+                            }
+                            cacheTile(totalCycleCount, cycleData[totalCycleCount].tileInfo, currentCycleData);
+                        }
+                    } catch (IOException e) {
+                        // when logging the error, increment cycle by 1, since totalCycleCount is zero-indexed but Illumina directories are 1-indexed.
+                        throw new PicardException(String.format("Error while reading from BCL file for cycle %d. Offending file on disk is %s",
+                                (totalCycleCount + 1), this.streamFiles[totalCycleCount].getAbsolutePath()), e);
+                    }
+                    totalCycleCount++;
+                }
+            }
 
         } catch (final IOException ioe) {
             throw new RuntimeIOException(ioe);
         }
     }
 
+    private Map<Integer, Map<Integer, File>> sortCbcls(List<File> cbcls) {
+        Map<Integer, Map<Integer, File>> sortedMap = new TreeMap<>();
+        Pattern pattern = Pattern.compile("^.+C(\\d{1,4}).+L(\\d{1,3})_(\\d).cbcl$");
+        for (File cbcl : cbcls) {
+            Matcher matcher = pattern.matcher(cbcl.getAbsolutePath());
+            if (!matcher.matches()) {
+                throw new PicardException("CBCL File " + cbcl.getAbsolutePath() + " does not match expected pattern.");
+            }
+            Integer surface = Integer.valueOf(matcher.group(3));
+            Integer cycle = Integer.valueOf(matcher.group(1));
+            if (sortedMap.containsKey(surface)) {
+                sortedMap.get(surface).put(cycle, cbcl);
+            } else {
+                Map<Integer, File> cycleMap = new HashMap<>();
+                cycleMap.put(cycle, cbcl);
+                sortedMap.put(surface, cycleMap);
+            }
+        }
+
+        return sortedMap;
+    }
+
     @Override
     public boolean hasNext() {
-        if (queue.isEmpty()) {
+        if (queue == null) {
             advance();
         }
-        return !queue.isEmpty();
+        return !(queue == null);
```

Comment:
Is this form preferred over `queue != null`? It seems to me that `!=` is clearer than `!(==)`